src/drivers/intel/dmic.c

```diff
@@ -1649,9 +1656,9 @@ static int dmic_remove(struct dai *dai)
 	interrupt_unregister(dmic->irq, dai);
 
 	/* The next end tasks must be passed if another DAI FIFO still runs */
-	dai_info(dai, "dmic_remove(), dmic_active_fifos_mask = 0x%x",
-		 uncached_dmic_active_fifos_mask);
-	if (uncached_dmic_active_fifos_mask)
+	dai_info(dai, "dmic_remove(), dmic_active_fifos_mask = 0x%x, dmic_pause_mask = 0x%x",
+		 uncached_dmic_active_fifos_mask, uncached_dmic_pause_mask);
+	if (uncached_dmic_active_fifos_mask || uncached_dmic_pause_mask)
 		return 0;
 
 	pm_runtime_put_sync(DMIC_CLK, dai->index);
```

Comment:
This is really problematic. There's no actual reference counting in cavs_pm_runtime_dis_dmic_clk_gating() and the "dai->index" passed to pm_runtime_put_sync() is just printed out.

The problem is that if we switch to pm_runtime_put_sync() that actual does refcounting, this code will break. FYI @lgirdwood @lyakh @marc-hb 

Oh well, this goes beyond this PR's scope.

Reply:
Lets get the ref counting in the PM code. This PM code should also be able to transfer to Zephyr with minimal refactoring.

Reply:
it looks like we may need some locking around the mask manipulations .

Reply:
I added 3rd commit about spinlocks. I was testing spinlock add in dmic_remove() but it froze. It turned out that dai_put() had already applied spinlock. I also found out that the registers write function was already covered by spinlock in set_config() so the patch adds only comments about spinlocks applied.